File: utilcode/lib/src/main/java/com/blankj/utilcode/util/AdaptScreenUtils.java
package com.blankj.utilcode.util;

import android.content.res.Resources;
import android.util.DisplayMetrics;
import android.util.Log;

import java.lang.reflect.Field;


public final class AdaptScreenUtils {

    private static boolean sIsInit;
    private static Field   sMetricsField;

    /**
     * Adapt for the horizontal screen, and call it in [android.app.Activity.getResources].
     */
    public static Resources adaptWidth(Resources resources, int designWidth) {
        DisplayMetrics dm = getDisplayMetrics(resources);
        float newXdpi = dm.xdpi = (dm.widthPixels * 72f) / designWidth;
        setAppDmXdpi(newXdpi);
        return resources;
    }

    /**
     * Adapt for the vertical screen, and call it in [android.app.Activity.getResources].
     */
    public static Resources adaptHeight(Resources resources, int designHeight) {
        DisplayMetrics dm = getDisplayMetrics(resources);
        float newXdpi = dm.xdpi = (dm.heightPixels * 72f) / designHeight;
        setAppDmXdpi(newXdpi);
        return resources;
    }

    /**
     * @param resources The resources.
     * @return the resource
     */
    public static Resources closeAdapt(Resources resources) {
        DisplayMetrics dm = getDisplayMetrics(resources);
        float newXdpi = dm.xdpi = dm.density * 72;
        setAppDmXdpi(newXdpi);
        return resources;
    }

    /**
     * Value of pt to value of px.
     *
     * @param ptValue The value of pt.
     * @return value of px
     */
    public static int pt2Px(float ptValue) {
        DisplayMetrics metrics = Utils.getApp().getResources().getDisplayMetrics();
        return (int) (ptValue * metrics.xdpi / 72f + 0.5);
    }

    /**
     * Value of px to value of pt.
     *
     * @param pxValue The value of px.
     * @return value of pt
     */
    public static int px2Pt(float pxValue) {
        DisplayMetrics metrics = Utils.getApp().getResources().getDisplayMetrics();
        return (int) (pxValue * 72 / metrics.xdpi + 0.5);
    }

    private static void setAppDmXdpi(final float xdpi) {
        Utils.getApp().getResources().getDisplayMetrics().xdpi = xdpi;
    }

    private static DisplayMetrics getDisplayMetrics(Resources resources) {
        DisplayMetrics realMetrics = getRealMetrics(resources);
        if (realMetrics != null) return realMetrics;
        return resources.getDisplayMetrics();
    }

    private static DisplayMetrics getRealMetrics(final Resources resources) {
        if (!sIsInit) {
            DisplayMetrics ret = null;
            Class resCls = resources.getClass();
            Field[] declaredFields = resCls.getDeclaredFields();
            while (ret == null && declaredFields != null && declaredFields.length > 0) {
                for (Field field : declaredFields) {
                    if (field.getType().isAssignableFrom(DisplayMetrics.class)) {
                        field.setAccessible(true);
                        sMetricsField = field;
                        ret = getMetricsFromField(resources);
                        if (ret != null) break;
                    }
                }
                resCls = resCls.getSuperclass();
                if (resCls != null) {
                    declaredFields = resCls.getDeclaredFields();
                } else {
                    break;
                }
            }
            sIsInit = true;
            return ret;
        }
        if (sMetricsField == null) return null;
        return getMetricsFromField(resources);
    }

    private static DisplayMetrics getMetricsFromField(final Resources resources) {
        try {
            return (DisplayMetrics) sMetricsField.get(resources);
        } catch (Exception e) {
            Log.e("AdaptScreenUtils", "getMetricsFromField: " + e);
            return null;
        }
    }
}

File: utilcode/src/main/java/com/blankj/utilcode/util/AdaptScreenUtils.kt
package com.blankj.utilcode.util

import android.content.Context
import android.content.res.Resources


class AdaptScreenUtils private constructor() {

    companion object {
        /**
         * Adapt for the horizontal screen, and call it in [android.app.Activity.getResources].
         */
        @JvmStatic
        fun adaptWidth(resources: Resources, designWidth: Int, isOpen: Boolean = true): Resources {
            val dm = resources.displayMetrics
            dm.xdpi = if (isOpen) (dm.widthPixels * 72f) / designWidth else dm.density * 72f
            return resources
        }

        /**
         * Adapt for the vertical screen, and call it in [android.app.Activity.getResources].
         */
        @JvmStatic
        fun adaptHeight(resources: Resources, designHeight: Int, isOpen: Boolean = true): Resources {
            val dm = resources.displayMetrics
            dm.xdpi = if (isOpen) (dm.heightPixels * 72f) / designHeight else dm.density * 72f
            return resources
        }

        /**
         * Value of pt to value of px.
         */
        @JvmStatic
        fun pt2Px(context: Context, value: Float): Float {
            val metrics = context.resources.displayMetrics
            return value * metrics.xdpi * (1.0f / 72);
        }

        /**
         * Value of px to value of pt.
         */
        @JvmStatic
        fun px2Pt(context: Context, value: Float): Float {
            val metrics = context.resources.displayMetrics
            return value * 72 / metrics.xdpi;
        }
    }
}